adafruit_register/i2c_struct_array.py

```diff
@@ -0,0 +1,108 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+# pylint: disable=too-few-public-methods
+
+"""`i2c_struct_array` - Array of structured registers based on `struct`
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+"""
+
+try:
+    import struct
+except ImportError:
+    import ustruct as struct
+
+__version__ = "0.0.0-auto.0"
+__repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_Register.git"
+
+class _BoundStructArray:
+    """
+    Actual array object that `StructArray` constructs on demand for
+
+    :param object obj: The device object to bind to. It must have a `i2c_device` attribute
+    :param int register_address: The register address to read the bit from
+    :param type struct_format: The struct format string for each register element
+    :param int count: Number of elements in the array
+    """
+    def __init__(self, obj, register_address, struct_format, count):
+        self.format = struct_format
+        self.first_register = register_address
+        self.obj = obj
+        self.count = count
+
+    def _header(self, index):
+        """Shared bounds checking and buffer creation."""
+        if index < 0 or index >= self.count:
+            raise IndexError()
+        size = struct.calcsize(self.format)
+        # We create the buffer every time instead of keeping the buffer (which is 32 bytes at least)
+        # around forever.
+        buf = bytearray(size + 1)
+        buf[0] = self.first_register + size * index
+        return buf
+
+    def __getitem__(self, index):
+        buf = self._header(index)
+        with self.obj.i2c_device:
+            self.obj.i2c_device.write(buf, end=1, stop=False)
+            self.obj.i2c_device.readinto(buf, start=1)
+        return struct.unpack_from(self.format, buf, offset=1)
+
+    def __setitem__(self, index, value):
+        buf = self._header(index)
+        struct.pack_into(self.format, buf, 1, *value)
+        with self.obj.i2c_device:
+            self.obj.i2c_device.write(buf)
+
+    def __len__(self):
+        return self.count
+
+class StructArray:
+    """
+    Repeated array of structured registers that are readable and writeable.
+
+    Based on the index, values are offset by the size of the structure.
+
+    Values are tuples that map to the values in the defined struct.  See struct
+    module documentation for struct format string and its possible value types.
+
+    :param int register_address: The register address to begin reading the array from
+    :param str struct_format: The struct format string for this register.
+    :param int count: Number of elements in the array
+    """
+    def __init__(self, register_address, struct_format, count):
+        self.format = struct_format
+        self.address = register_address
+        self.count = count
+        self.array_id = "_structarray" + str(register_address)
```

Comment:
`.format()` would be more readable here.

Reply:
Done